sentinel-adapter/sentinel-zuul2-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/gateway/zuul2/filters/inbound/SentinelZuulInboundFilter.java

```diff
@@ -0,0 +1,151 @@
+
+    private final int order;
+
+    private final String blockedEndpointName;
+    /**
+     * if executor is null, flow control action will do on I/O thread
+     */
+    private final Executor executor;
+
+    private final GatewayParamParser<HttpRequestMessage> paramParser = new GatewayParamParser<>(
+            new HttpRequestMessageItemParser());
+
+    public SentinelZuulInboundFilter(int order) {
+        this(order, null);
+    }
+
+	public SentinelZuulInboundFilter(int order, Executor executor) {
+        this(order, DEFAULT_BLOCK_ENDPOINT_NAME, executor);
+	}
+
+    public SentinelZuulInboundFilter(int order, String blockedEndpointName, Executor executor) {
+        this.order = order;
+		this.blockedEndpointName = blockedEndpointName;
+		this.executor = executor;
+	}
+
+    @Override
+    public int filterOrder() {
+        return order;
+    }
+
+    @Override
+    public Observable<HttpRequestMessage> applyAsync(HttpRequestMessage request) {
+        if (executor != null) {
+            return Observable.just(request).subscribeOn(Schedulers.from(executor)).flatMap(this::apply);
+        } else {
+            return Observable.just(request).flatMap(this::apply);
+        }
+    }
+
+    private Observable<HttpRequestMessage> apply(HttpRequestMessage request) {
+        SessionContext context = request.getContext();
+        String origin = parseOrigin(request);
+        Deque<AsyncEntry> asyncEntries = new ArrayDeque<>();
+        String routeId = (String) context.get(ZuulConstant.PROXY_ID_KEY);
+        String fallBackRoute = routeId;
+        try {
+            if (StringUtil.isNotBlank(routeId)) {
+                ContextUtil.enter(GATEWAY_CONTEXT_ROUTE_PREFIX + routeId, origin);
+                doSentinelEntry(routeId, RESOURCE_MODE_ROUTE_ID, request, asyncEntries);
+            }
+            Set<String> matchingApis = pickMatchingApiDefinitions(request);
+            if (!matchingApis.isEmpty() && ContextUtil.getContext() == null) {
+                ContextUtil.enter(ZuulConstant.ZUUL_DEFAULT_CONTEXT, origin);
+            }
+            for (String apiName : matchingApis) {
+                fallBackRoute = apiName;
+                doSentinelEntry(apiName, RESOURCE_MODE_CUSTOM_API_NAME, request, asyncEntries);
+            }
+            return Observable.just(request);
+        } catch (BlockException t) {
+			context.put(ZuulConstant.ZUUL_CTX_SENTINEL_BLOCKED_FLAG, Boolean.TRUE);
+			context.put(ZuulConstant.ZUUL_CTX_SENTINEL_FALLBACK_ROUTE, fallBackRoute);
+            context.setEndpoint(blockedEndpointName);
+            return Observable.error(t);
+        } finally {
+            if (!asyncEntries.isEmpty()) {
+                context.put(ZuulConstant.ZUUL_CTX_SENTINEL_ENTRIES_KEY, asyncEntries);
+                // clear context to avoid another request use incorrect context
+                ContextUtil.exit();
```

Comment:
need to deliver the context to "SessionContext" and exit the context in outbound filter.There is a limit in [ContextUtil](https://github.com/alibaba/Sentinel/blob/ff5c010a514d612a12fd49693f7a8c8a6e717dee/sentinel-core/src/main/java/com/alibaba/csp/sentinel/context/ContextUtil.java#L126),and there will be memory leak if not clean correctly.

Reply:
threadlocal of context may will not clean up in your case, i will fix it. And i think the place of `ContextUtil.exit()` is not important in event-driven model, it's function is  just to clean threadlocal,  but it maybe necessary to store the context for other purpose instead of discarding it.